### ISSUE TITLE
sqlalchemy need to relax certain server capacities assumption and better handle the errors to work with more databases

### ISSUE THREAD

**Issue by LLuke** [sqlalchemy/sqlalchemy#10656]
### Describe the use case

A fix will allow sqlalchemy to work with some database servers implemented enough postgres wire protocol but not all the PostgreSQL features assumed by sqlalchemy.

### Description
sqlalchemyattempted to use queries in 

```
 engine.default.get_server_version_info()
 engine.default.._get_default_schema_name()
 engine.default..get_default_isolation_level()
```
and
```
 dialects.postgresql.PGDialect.self._set_backslash_escapes()
```

to probe and enable features for PostgreSQL and compatible databases. While there are good reasons, they could fail with unhandled errors for database servers that support enough postgres wire protocol and features but otherwise doesn't support these particular functionalities. In particular, sqlalchemy assumed the error would always be NotImplementedError in get_server_version_info() and proceeded to use  server_version_info without further check the None condition arise from NotImplementedError. However, for drivers like pg8000, errors like ProgrammingError would be thrown instead. To make sqlalchemy usable for the many different drivers and servers combinations, it is desirable to make sqlalchemy as lenient as possible without detrimental side effects.

### Databases / Backends / Drivers targeted

A fix is tested with arcadedb 23.10.1 with pg8000. It is possible this will also fix for other servers implemented enough postgres wire protocols

### Example Use

After the fix, the following code works
```
from sqlalchemy import create_engine
import pandas as pd

engine = create_engine(
    'postgresql+pg8000://' + os.environ['DB_USER'] + ':' + os.environ['DB_PASS'] + '@localhost:3333/arcade_testdb', isolation_level="AUTOCOMMIT",
    connect_args={'ssl_context': None}
)
df = pd.read_sql_query("select * from schema:database", con=engine)
engine.dispose()
```

**Comment by CaselIT** [comment]
Hi,

In general sqlalchemy tries to avoid having to support the many flavours and combinations of database that kinda but not really support the postgresql protocol.

The reason is that claiming to do so will lead to a endless stream of issues were people try to use a particular feature that's not implemented or works in a subtly different way.

The usual solution is to create an external dialect for it, that can be based on the internal pg one.

Let's wait for Mike opinion, but personally I'm -1 on this

**Comment by zzzeek** [comment]
hi - 

for examples of PostgreSQL variant dialects, see https://github.com/sqlalchemy-redshift/sqlalchemy-redshift  , https://github.com/cockroachdb/sqlalchemy-cockroachdb , and probably others.

The SQLAlchemy project can't write code that can't be tested locally; if we were to do something like this:

```py
try:
      _do_thing()
except:
      pass
```

instead of expecting the error that is "checked":

```py
try:
      _do_thing()
except NotImplementedError:
      pass
```

we can't test that code.  We dont have any dialect that raises an error besides `NotImplementedError`, so that remains an uncovered path.    if some dialect did raise an exception that's not `NotImplementedError`, it could be somehting like `TypeError`, or `OperationalError('database not connected')`, or anything else, that is an error we've never anticipated and that nobody else has ever anticipated, and therefore it needs to raise.  it's undefined behavior, and allowing undefined behavior to pass silently is a bug.

basically it's not possible to write and maintain code that expects the implementation to be broken, since "broken" can mean anything.

closing this as "wontfix" as you will need to create a small dialect of your own, that just does the "server_version_info" part here, that's quite simple.   Background on custom dialect is at https://github.com/sqlalchemy/sqlalchemy/blob/main/README.dialects.rst

**Comment by LLuke** [comment]
Agree in principle and guidance on potential solutions, also agree on always catch typed error rather than the blank catch all.

Just nitpicking though in theory sqlalchemy cannot rely on NotImplementedError since one project's mandate is not necessarily in another project's contract. In my testing, I found that pg8000 throws ProgrammingError since they consider failing of any query execution the same. Their understanding of NotImplementedError is obviously different from sqlalchemy 's. In in such case, sqlalchemy should still have a final catch all, log and translate the known "not implemented" scenarios into NotImplementedError for its own logic. This for sure is moot and such nuanced approach will not be needed if PostgreSQL always implements the assumed functionalities. My first very simple shot on this is as a third-party user to hear your thoughts and pass the information back on to the projects so that they can consider.

Have a great week!

> hi -
> 
> for examples of PostgreSQL variant dialects, see https://github.com/sqlalchemy-redshift/sqlalchemy-redshift , https://github.com/cockroachdb/sqlalchemy-cockroachdb , and probably others.
> 
> The SQLAlchemy project can't write code that can't be tested locally; if we were to do something like this:
> 
> ```python
> try:
>       _do_thing()
> except:
>       pass
> ```
> 
> instead of expecting the error that is "checked":
> 
> ```python
> try:
>       _do_thing()
> except NotImplementedError:
>       pass
> ```
> 
> we can't test that code. We dont have any dialect that raises an error besides `NotImplementedError`, so that remains an uncovered path. if some dialect did raise an exception that's not `NotImplementedError`, it could be somehting like `TypeError`, or `OperationalError('database not connected')`, or anything else, that is an error we've never anticipated and that nobody else has ever anticipated, and therefore it needs to raise. it's undefined behavior, and allowing undefined behavior to pass silently is a bug.
> 
> basically it's not possible to write and maintain code that expects the implementation to be broken, since "broken" can mean anything.
> 
> closing this as "wontfix" as you will need to create a small dialect of your own, that just does the "server_version_info" part here, that's quite simple. Background on custom dialect is at https://github.com/sqlalchemy/sqlalchemy/blob/main/README.dialects.rst

**Comment by zzzeek** [comment]
> Agree in principle and guidance on potential solutions, also agree on always catch typed error rather than the blank catch all.
> 
> Just nitpicking though in theory sqlalchemy cannot rely on NotImplementedError since one project's mandate is not necessarily in another project's contract. In my testing, I found that pg8000 throws ProgrammingError since they consider failing of any query execution the same. 

the blocks that catch NotImplementedError here are SQLAlchemy dialect methods and they have a contract of throwing NotImplementedError if the dialect does not support the operation.   if the dialect does support the operation, but then it fails for some other reason, that's a different codepath, that propagates out.

so there is no issue with DBAPIs throwing ProgrammingError regarding these methods.   if the DBAPI throws a ProgrammingError and the SQLAlchemy dialect method is not ready for that, that's called an "unchecked exception" and it's an error.

> Their understanding of NotImplementedError is obviously different from sqlalchemy 's. In in such case, sqlalchemy should still have a final catch all, log and translate the known "not implemented" scenarios into NotImplementedError for its own logic.

no, because if the DBAPI is not connecting, is not compatible, or is experiencing a bug of some kind, that's undefined behavior.  the program has to stop, otherwise there will be more errors later that may be more difficult to debug.  there's lots of videos and blog posts about this, this one seems to cover it among others: https://jerrynsh.com/stop-using-exceptions-like-this-in-python/

**Comment by LLuke** [comment]
Even though I learned that all assumptions and fixes have their tradeoffs in the decades, for sure I understand your point of view and really appreciate your generosity in sharing all of your good ideas . 👍
